runner/bsp/worker/src/mill/bsp/worker/MillBuildServer.scala

```diff
@@ -660,83 +661,126 @@ private class MillBuildServer(
     }
   }
 
-  val requestLock = new java.util.concurrent.locks.ReentrantLock()
+  private val queue = new LinkedBlockingQueue[(BspEvaluators => Unit, Logger, String)]
+  private var stopped = false
+  private val evaluatorRequestsThread: Thread =
+    new Thread("mill-bsp-evaluator") {
+      setDaemon(true)
+      def waitForEvaluators(): BspEvaluators =
+        bspEvaluatorsOpt match {
+          case Some(ev) => ev
+          case None =>
+            Thread.sleep(1000L)
```

Comment:
Each `Thead.sleep` should come with a comment: Why and why the value?

Reply:
The last push gets rid of it, by using a promise and awaiting it to have a value there